Add Annotation Search and Verification Pages

hudson-newey opened the pull request:
# Verification Interface

This pull request creates an annotations search page & accompanying verification grid to the workbench client.

## Changes

### Features

- Add open-ecoacoustics web components as a dependency
- Add annotations search page
- Add verification interface page
- Add `@angular/elements` allowing us to create custom elements using Angular components (was needed for full Lit + Angular interoperability within slotted content)
- Add semantic selectors to wip component
- Add disabled attributes to `date-time-input.component` start/end date/time inputs
- Add `pageSize` attribute to pagination template
- You can now add response headers to the SSR server using the `angular.json` `serve.options.headers` object

### Bug Fixes

- Fix the previously non-functional `UnsavedInputGuard`
- Fix docker image publish workflow doubling the inputted tag name on manual dispatches
- Fix docker version build numbers now use an incrementing number instead of the number of commits in the tag
- Fixes a bug with route titles when viewing profiles
- Fixes a bug where typeahead input component pills would not be vertically centered
- Fix failing CI due to deprecated dependencies

### Code Quality

- Disable AOT during development builds and disable lib check (don't typecheck third party library code). These changes result in decreased dev build time by ~40-50%
- Add new test helpers `selectFromTypeahead` and `getElementByInnerText`

## Issues

Fixes: #2140 

## Visual Changes

![image](https://github.com/user-attachments/assets/dc8ddb8b-84eb-4f81-858e-5a57fab233fa)

_Annotations search page_

---

![image](https://github.com/user-attachments/assets/cccbc26e-b707-42c1-827a-19e1678117b0)

_Verification grid page_

---

![image](https://github.com/user-attachments/assets/6febe2e4-5b7d-4781-bf20-1c3f245db820)

_Verification grid with context card expanded_

## Final Checklist

- [x] Assign reviewers if you have permission
- [x] Assign labels if you have permission
- [x] Link issues related to PR
- [ ] Ensure project linter is not producing any warnings (`npm run lint`)
- [ ] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

hudson-newey committed:
Created verification interface

hudson-newey committed:
Abstract out verification form

hudson-newey committed:
Backfilled verification models

hudson-newey committed:
Created rudimentary search page

hudson-newey committed:
Pushing

hudson-newey committed:
Minor changes to verification interface

hudson-newey committed:
polish

hudson-newey committed:
Push

hudson-newey committed:
Add tests

hudson-newey committed:
Update verification component build

hudson-newey committed:
Correctly import web components

hudson-newey committed:
Add auth token to verification grid media requests

hudson-newey committed:
Basic test boilerplate

hudson-newey committed:
Consolidate search and verification grid page

hudson-newey committed:
Basic unit test outline

hudson-newey committed:
Update web-components version

hudson-newey committed:
Add outline for web component tests

hudson-newey committed:
Add warning modal for lost progress

hudson-newey committed:
Improve shadow root selectors

hudson-newey committed:
Restructure annotation routes

hudson-newey committed:
Add test fixtures for new routes

hudson-newey committed:
Improve test helpers

hudson-newey committed:
Staging annotation search form tests

hudson-newey committed:
Add id field to annotation search form

hudson-newey committed:
Staging to use pagination template

hudson-newey committed:
Use paginationTemplate for search results page

hudson-newey committed:
Display warning when there are no filter conditions

hudson-newey committed:
Add context to grid tiles

hudson-newey committed:
Add support for UnsavedInputGuard

hudson-newey committed:
Add theming colors to web components

hudson-newey committed:
Working mvp of verification interface

hudson-newey committed:
Fix search parameters on search page

hudson-newey committed:
Do not show QSP for page=1

hudson-newey committed:
Minor improvements

hudson-newey committed:
Add support for adding custom elements

hudson-newey committed:
Fix grid tile content component being rendered twice

hudson-newey committed:
Feedback changes

- Context card now uses anchor positioning
- Add timeout before scroll to verification grid
- Verification grid route now has annotation search in breakcrumbs
- Upgrade npm version of web components
- Add media controls to context popup
- Tile slotted content now uses button instead of link for context
- Filter icon is now smaller with padding

hudson-newey committed:
Fix pagination in verification grid

hudson-newey committed:
Add grid tile content tests

hudson-newey committed:
Add context request tests to grid tile content component

hudson-newey committed:
Add tests for audio event card component

hudson-newey committed:
Fix context popup overflow

hudson-newey committed:
Compact tile slotted content

hudson-newey committed:
Remove verification models and services

hudson-newey committed:
Ann annotation data model

hudson-newey committed:
Update outdated CI dependencies

hudson-newey committed:
Hack: Increase bundle size

hudson-newey committed:
Fix ssr rendering

hudson-newey committed:
Decrease bundle size & dynamically import context token

hudson-newey committed:
Remove mock dataset

hudson-newey committed:
Staging changes

atruskie commented on src/app/app.component.ts:
```diff
@@ -83,6 +100,8 @@ export class AppComponent extends withUnsubscribe() implements OnInit {
       return;
     }
 
+    import("@ecoacoustics/web-components");
```
probably need to explain why this import is dynamic

atruskie commented on src/app/app.component.ts:
```diff
@@ -64,6 +71,16 @@ export class AppComponent extends withUnsubscribe() implements OnInit {
     this.router.initialNavigation();
     globals.initialize();
+
+    if (!this.isServer) {
```
comment

```
// register all web components here
// we make some of our standalone angular components into standards based web components
// so that they can operate entirely independently - e.g. in shadow dom
```

atruskie commented on src/app/components/annotations/annotation.menu.ts:
```diff
@@ -0,0 +1,100 @@
+    icon: ["fas", "layer-group"],
+    label: "Search Annotations",
+    tooltip: () => "(BETA) Search Annotations",
+    predicate: isLoggedInPredicate,
```
showing annotations should be available on public anonymous access for public access projects

atruskie commented on src/app/components/annotations/components/annotation-search-form/annotation-search-form.component.html:
```diff
@@ -0,0 +1,94 @@
+    (modelChange)="
+      updateModel('audioRecordings', getIdsFromAbstractModelArray($event))
+    "
+  ></baw-typeahead-input>
```
ok this is an id search. this is a pretty advanced filter and not that friendly to use. Honestly most of the time if this will be used it will be activated by an external cold load from a generated url.

- move it the bottom of the search parameters (least important)
- maybe hide it in an advanced section
- and the label need to be clearer - it's an ID filter but that's not in the name

hudson-newey replied:
Mock up with "Advanced Filters" section

![image](https://github.com/user-attachments/assets/031b1868-0cb4-4758-9dde-dedc1a8276da)

hudson-newey replied:
Feedback from offline review: This should be an accordian not a collapsable

hudson-newey replied:
Update advanced filters section

![image](https://github.com/user-attachments/assets/a53dbc84-0ab6-4af4-bf59-43fabde59d4d)

Remember that WIP will be replaced with a placeholder when deployed to production

atruskie commented on src/app/components/annotations/components/annotation-search-form/annotation-search-form.component.html:
```diff
@@ -0,0 +1,94 @@
+      Only show un-verified annotations
+    </label>
+  </div>
+
```
wrap in a `wip` component so it is hidden

atruskie commented on src/app/components/annotations/components/annotation-search-form/annotation-search-form.component.ts:
```diff
@@ -0,0 +1,134 @@
+        dateTimeModel.timeFinishedBefore,
+      ];
+    }
+  }
```
it feels like this code must have existed elsewhere. Is this a copy? Can we reuse the code? A bunch of static helpers?

the same goes for everything else in this file

atruskie commented on src/app/components/annotations/components/annotation-search-form/annotation-search-form.component.ts:
```diff
@@ -0,0 +1,134 @@
+            text as any,
+            includeDefaultFilters && this.defaultFilter()
+          ),
+          notIn<T>(key as keyof AbstractModel, activeItems)
```
i'm going to need a comment for this one. 

oh i think i worked it out: for a new typeahead query, only suggest things that haven't already been chosen

atruskie commented on src/app/components/annotations/components/broad-filters-warning/broad-filters-warning.component.ts:
```diff
@@ -0,0 +1,50 @@
+    <div class="modal-body">
+      <p>
+        You no applied filter conditions. The verification grid will display all
+        available tasks.
```
```suggestion
        You are seeing <i>all</i> annotations because no search filters were used.
        All annotations that will be shown for verification.
```

atruskie replied:
future feature request: show number of annotations that will be in the set for the all/no filter case. Mybe omit the warning if it is less than 1000 etc..

atruskie commented on src/app/components/annotations/components/broad-filters-warning/broad-filters-warning.component.ts:
```diff
@@ -0,0 +1,50 @@
+
+      <p>
+        This can result in a verification grid that is difficult to navigate and
+        may take a long time to complete.
```
```suggestion
        You can verify all annotations in one go, but it may take a long time to complete
        and it will be difficult to navigate. We suggest breaking up the task into smaller
        chunks.
```

atruskie commented on src/app/components/annotations/components/grid-tile-content/grid-tile-content.component.html:
```diff
@@ -0,0 +1,44 @@
+      </oe-indicator>
+    </oe-axes>
+
+    <!-- <oe-media-controls [for]="contextSpectrogramId"></oe-media-controls> -->
```
why disabled?

hudson-newey replied:
Angular `[for]` attribtue gets removed when used on this media controls (for some reason)

I've commented it out for the time being, so that demos don't have a broken media controls component

You'll see that I've got the media controls working in the search card component using `ViewChild`. That's a hack, and not ideal.

I couldn't use `ViewChild` here (for some reason). Perhaps because it's in the shadow dom?

atruskie commented on src/app/components/annotations/components/grid-tile-content/grid-tile-content.component.html:
i'd like to establish a naming convention for web-components

since they're standalone, they're truly different from other components we have and i'd like to demarcate that clearly

suggestions:

`web-components` folder?
`web-component` in the extension?

one or both or something else?

hudson-newey replied:
I like the idea of a `web-components` folder

The Angular style guide is pretty opinionated on file naming conventions, and I don't want to stray to far away from them.

Surprisingly, Angular doesn't specify any file structure for web components, so I think a web-components folder breaks the least Angular assumptions

If we were to put `web-component` in the file extension I think the least objectable option would be something like `grid-tile-content.web-component.component.ts` and at the worst case `grid-tile-content.web-component.component.spec.ts`

atruskie replied:
we agree on the folder

definite no to a triple extension.

I'm going to executively say either:

```
./web-components/a.component.ts
```

or 

```
./web-components/a.web-component.ts
```

you can choose

atruskie commented on src/app/components/annotations/components/grid-tile-content/grid-tile-content.component.html:
```diff
@@ -0,0 +1,44 @@
+        <oe-spectrogram
+          [id]="contextSpectrogramId"
+          [src]="contextSource"
+          color-map="audacity"
```
we should use grayscale here to match our host site

todo later: spectrogram settings per user site wide

atruskie commented on src/app/components/annotations/components/grid-tile-content/grid-tile-content.component.ts:
```diff
@@ -0,0 +1,114 @@
+    newUrl.search = searchParams.toString();
+
+    return newUrl.toString();
+  }
```
this should be moved out of this component (TODO) and into the model component that's passed into it so we're just calling `model.contextUrl(30)` or something like that

atruskie commented on src/app/components/annotations/components/grid-tile-content/grid-tile-content.component.ts:
```diff
@@ -0,0 +1,114 @@
+        this.handleContextChange.bind(this),
+        true
+      )
+    );
```
we should bind this up into a decorator on a field just like `@ViewElementRef` (TODO)

hudson-newey replied:
Something like this seems to do the trick

![image](https://github.com/user-attachments/assets/8df8914e-40fc-44bd-8684-ce5b70223945)

atruskie commented on src/app/components/annotations/components/grid-tile-content/grid-tile-content.component.ts:
```diff
@@ -0,0 +1,114 @@
+  }
+
+  public handleContextChange(subjectWrapper: SubjectWrapper): void {
+    this.model.set(new Annotation(subjectWrapper.subject));
```
why are we `new`ing here? shouldn't this be an `as` - just a type cast?

hudson-newey replied:
Good catch, the constructor prototype will still be attached because the web components pass around the subject by reference (we did this change a while back so that we would download results more efficiently)

atruskie commented on src/app/components/annotations/components/reset-progress-warning/reset-progress-warning.component.ts:
```diff
@@ -0,0 +1,53 @@
+    <div class="modal-body">
+      <p>
+        Because you have changed the verification tasks search parameters, your
+        progress will be reset if you continue.
```
```suggestion
        Because you have changed the verification search parameters, your
        progress will be reset if you continue.
```

atruskie commented on src/app/components/annotations/components/reset-progress-warning/reset-progress-warning.component.ts:
```diff
@@ -0,0 +1,53 @@
+
+      <p>
+        Do you wish to reset your progress and continue with the new search
+        parameters?
```
```suggestion
        Do you wish to reset your progress and continue with the new search
        parameters?
        <br/>
        <!-- TODO: replace with don't worry everything have been saved when we implement server persistence -->
        If you do wish to continue, make sure you download the annotations you
        have already done first!
```

atruskie commented on src/app/components/annotations/pages/search/search.component.html:
```diff
@@ -0,0 +1,72 @@
+  <div class="mt-2">
+    <!--
+      TODO: Once the audio events download route supports url encoded filters
+      we should uncomment this buttons functionality
```
file an api issue and link back

hudson-newey replied:
Blocked by https://github.com/QutEcoacoustics/baw-server/issues/688

atruskie commented on src/app/components/annotations/pages/annotationSearchParameters.ts:
```diff
@@ -0,0 +1,156 @@
+    return filterModelIds("projects", this.projects);
+  }
+
+  //! FOR PR REVIEWER: THIS IS WHERE I AM CURRENTLY UP TO
```
expand into todo with context

atruskie commented on src/app/helpers/filters/audioEventFilters.ts:
```diff
@@ -0,0 +1,41 @@
+  startTime?: Duration,
+  endTime?: Duration
+): InnerFilter<AudioEvent> {
+}
```
as we discussed, not available with current API. leave these stubs and raise not implemented. cross link with api issue

hudson-newey committed:
Quick hacky fix for project scoping

The correct fix here will be to use route resolvers and add the project,
region, and site filters to the annotation search parameters data model.

This is because I cannot directly use the association decorators because
they are async, and do not await. Therefore returning before the full
model can be fetched.

atruskie commented on src/app/models/data/Annotation.ts:
```diff
@@ -0,0 +1,49 @@
+  audioLink: string;
+}
+
+export class Annotation extends AbstractModelWithoutId implements IAnnotation {
```
explain context behind this class

viewmodel

multiple models included to make rendering easier

multiple services used in a meta service to retrieve and create one of these

atruskie commented on src/app/services/models/annotation.service.ts:
```diff
@@ -0,0 +1,74 @@
+    @Inject(API_ROOT) private apiRoot: string
+  ) {}
+
+  public async buildAnnotation(audioEvent: AudioEvent): Promise<Annotation> {
```
todo: i'd like to see this class get an API like our other services. 

`show`, `create`, and `update`

atruskie commented on src/app/services/models/annotation.service.ts:
```diff
@@ -0,0 +1,74 @@
+    );
+  }
+
+  private async createAudioRecording(
```
`create` naming is problematic, especially in a service context

suggest `fetch` or `show` or `list`

atruskie commented on src/app/services/models/annotation.service.ts:
```diff
@@ -0,0 +1,74 @@
+      urlParams += `&user_id=${this.session.authToken}`;
+    }
+
+    return this.apiRoot + basePath + urlParams;
```
this belongs elsewhere - probably in our media service?

hudson-newey committed:
Warning dialog copy changes

atruskie reviewed (changes requested):
great work, a lot more to do but if we're functional let's do these things as needed

hudson-newey committed:
Copy feedback and fix auth token qsp

hudson-newey committed:
Patch tag array referencing issue

hudson-newey committed:
Various bug fixes and web components bump

atruskie commented on src/app/components/annotations/components/broad-filters-warning/broad-filters-warning.component.ts:
```diff
@@ -17,13 +17,19 @@ import { NgbActiveModal } from "@ng-bootstrap/ng-bootstrap";
-        You no applied filter conditions. The verification grid will display all
-        available tasks.
+        You are seeing <em>all</em> annotations because no search filters were
+        used. All annotations will be shown.
```
```suggestion
        used.
```

atruskie commented on src/app/components/annotations/components/broad-filters-warning/broad-filters-warning.component.ts:
```diff
@@ -17,13 +17,19 @@ import { NgbActiveModal } from "@ng-bootstrap/ng-bootstrap";
+      <p>
+        We suggest breaking up the task into smaller chunks. This can result in
+        a verification grid that is difficult to navigate and may take a long
+        time to complete.
```
```suggestion
        We suggest breaking up the task into smaller chunks. 
```

atruskie commented on src/app/components/annotations/pages/annotationSearchParameters.ts:
```diff
@@ -45,13 +44,12 @@ export interface IAnnotationSearchParameters {
   time: MonoTuple<Duration, 2>;
 }
 
+// we exclude project, region, and site from the serialization table because
+// we do not want them emitted in the query string
```
we don't?

seems like a bug waiting to happen

I'm inferring that we take these values from route context?

hudson-newey replied:
Will be fixed when I resolve

https://github.com/QutEcoacoustics/workbench-client/pull/2145#discussion_r1802117157

atruskie commented on src/app/components/annotations/pages/search/search.component.html:
```diff
@@ -18,25 +18,43 @@ <h1>Annotation Search</h1>
+        session.isLoggedIn
+          ? 'Verify the annotations in the list'
+          : 'You must be logged in to verify annotations'
+      "
```
this seems like it should be a reusable component (disabled unless logged in and tooltip to explain it)

hudson-newey replied:
Maybe a directive that we can add to buttons and similar elements (instead of a wrapper component).

e.g.

```html
<button [requireLoggedIn]></button>

<div [requireLoggedIn]></div>
```

_Edit: After trying it, it'll probably be easier with a reusable component because we have to wrap the content in a span. We could do that with a directive, but it'd be overcomplicated_

atruskie commented on src/app/services/models/annotation.service.ts:
```diff
@@ -41,34 +45,103 @@ export class AnnotationService {
+    return {
+      startTimeSeconds: proposedStartTime,
+      endTimeSeconds: proposedEndTime,
+    };
```
this function does a lot of the same calculations as the audio link function. I'd suggest inlining it.

hudson-newey replied:
Fixed and moved to `MediaService`

atruskie commented on src/app/services/models/annotation.service.ts:
```diff
@@ -41,34 +45,103 @@ export class AnnotationService {
+    const basePath =
+      `/audio_recordings/${audioEvent.audioRecordingId}/media.flac`;
+    let urlParams = `?audio_event_id=${audioEvent.id}` +
+      `&start_offset=${safeEventStartTime}&end_offset=${safeEventEndTime}`;
```
you've got most of the parts in place here but I'd suggest a few tweaks:

- accept padding as an argument (default 0)
- calculate midpoint first
- if duration < 1 && padding  == 0 
    - then set padding to (1 - duration)/2
- then pad
- then floor/ceil start/end
- then bound check
   - on over/under flow, take excess and add it to the other side

hudson-newey committed:
Fix publish workflow

hudson-newey committed:
Add advanced filters section to annotation search form

hudson-newey committed:
Fix unresolved model throwing errors

atruskie commented on src/app/components/annotations/pages/search/search.component.ts:
```diff
@@ -176,6 +179,29 @@ class AnnotationSearchComponent
+    } as any);
+
+    return initialFilter;
+  }
```
this exact same function is in `verification.component.ts`?

hudson-newey replied:
Yeah, I need to move this to `AnnotationSearchParameters.ts`.

To get the (resolved) project models in the annotation search parameters data model, I will need to use a resolver, which will take some time to make.

I did this hack so that I could get a working demo deployed to Ecosounds sooner, instead of spending a lot of time getting route resolvers working for the parameter data model

hudson-newey committed:
Fix review comments to start on qsp resolvers

hudson-newey committed:
Add annotation qsp model to resolver

hudson-newey committed:
Fix edge cases for media rounding

hudson-newey committed:
Only show warning when there are more than 1000 items

hudson-newey committed:
Automatically close context when paging

hudson-newey committed:
Grid tile component changes

hudson-newey committed:
typeahead callback factory methods and context decorators

hudson-newey committed:
Improve typing for context decorator

hudson-newey committed:
Move route models to annotationSearchParameters

hudson-newey committed:
Fix audio event card tests

hudson-newey committed:
Add disabled attributes to date/time filter component

hudson-newey committed:
Half way through self-review

hudson-newey committed:
Finish new service tests

hudson-newey committed:
Staging

hudson-newey committed:
Improve if-logged-in component and tests

hudson-newey committed:
Finish search page tests

hudson-newey committed:
Add verification annotation search form tests

hudson-newey committed:
9 tests remaining for verification component tests

hudson-newey committed:
Backfill verification component tests

hudson-newey committed:
More resiliant karama test runner

hudson-newey committed:
Prefill end date from url parameters

hudson-newey committed:
Fix some faulty annotation search form tests

hudson-newey committed:
Staging to test bundling changes

hudson-newey committed:
Fix broken bundling of workers & worklets in tests

hudson-newey committed:
Fix iframe and fetch errors in tests

hudson-newey committed:
Staging for esm compliant web component testing

hudson-newey committed:
Fix re-declaration of custom elements

hudson-newey committed:
Fix advanced filter toggling logic and leaky tests

hudson-newey committed:
Bump components version

hudson-newey committed:
Bump components version to fix aliases

hudson-newey committed:
Revert timeout

hudson-newey committed:
Fix CI conflicts and lint errors

hudson-newey committed:
Increase timeout for dropdown selectors

hudson-newey committed:
Fix race conditions by bumping web components

hudson-newey committed:
Don't rely on bootstrap auto-selection in dropdowns

hudson-newey committed:
Tempoarily comment out flakey test

hudson-newey committed:
Add APP_INITIALIZER to prevent race conditions in imports

hudson-newey committed:
Staging

atruskie commented on src/app/components/annotations/pages/annotationSearchParameters.ts:
```diff
@@ -99,38 +156,98 @@ export class AnnotationSearchParameters
+  public routeFilters(): InnerFilter<AudioEvent> {
+    let siteIds: number[] = [];
+
+    if (this.routeSiteId) {
+      siteIds = [this.routeSiteId];
+    } else if (this.routeRegionId) {
+      siteIds = Array.from(this.routeRegionModel.siteIds);
+    } else {
+      siteIds = Array.from(this.routeProjectModel.siteIds);
+    }
```
it would be better if these actually just filtered by the entities that are restricted rather than by site ids.

- it avoids the additional request
- and it's more semantically correct.

audio recording supports filtering by site, region, and project

hudson-newey replied:
Audio recordings support filtering by associated site, region, and project models.

Because this data model constructs filter conditions for the audio_events API, our filter condition would have to look something like

```jsonc
// POST /audio_events/filter
{
	"filters": {
		// notice the double "." on associated models
		"audio_recordings.projects.id": {
			"in": [1, 2, 3, 4]
		}
	}
}
```

However because this is a filter condition on an association of an association it seems to fail

![image](https://github.com/user-attachments/assets/4c8edc13-e06a-4a14-91db-4c262018d355)

---

Potentially resolved by

https://github.com/QutEcoacoustics/baw-server/issues/689

atruskie replied:
file a separate issue and cross link

hudson-newey replied:
Associated issue

https://github.com/QutEcoacoustics/baw-server/issues/690

atruskie commented on src/app/components/annotations/pages/annotationSearchParameters.ts:
```diff
@@ -99,38 +156,98 @@ export class AnnotationSearchParameters
+      this.recordingDateStartedAfter,
+      this.recordingDateFinishedBefore
     );
+    // const dateTimeFilter = filterTime(
+    //   dateFilter,
+    //   this.daylightSavings,
+    //   this.recordingTimeStartedAfter,
+    //   this.recordingTimeFinishedBefore
+    // );
```
why commented out?

hudson-newey replied:
```ts
// time filtering is currently disabled until we can filter on custom fields
// and association
// see https://github.com/QutEcoacoustics/baw-server/issues/689
// TODO: enable time filtering once the api adds support
```

![image](https://github.com/user-attachments/assets/4f240762-08b7-4920-a41e-888a6a976ae0)

atruskie commented on src/app/components/annotations/pages/search/search.component.ts:
```diff
@@ -144,13 +139,28 @@ class AnnotationSearchComponent
+      const responseMetadata = response[0].getMetadata();
+      const numberOfItems = responseMetadata.paging.total;
+
+      if (numberOfItems > itemWarningThreshold) {
+        const warningModal = this.modals.open(this.broadFilterWarningModal);
```
could we pass in the number of search results into the modal template? show it to the user

hudson-newey replied:
Done

![image](https://github.com/user-attachments/assets/766f8788-e67f-4efc-acce-75eb76fb74be)

atruskie commented on src/app/components/annotations/pages/verification/verification.component.ts:
```diff
@@ -129,7 +120,17 @@ class VerificationComponent
+    // without this automatic focusing, the user would have to click on the
+    // verification grid (e.g. to make a sub-selection) before being able to
+    // use the shortcut keys
+    this.verificationGridElement.nativeElement.focus();
```
i wonder if we need a focus prompt/indicator on the verification grid... it could be confusing if keyboard shortcuts suddenly don't work

hudson-newey replied:
![image](https://github.com/user-attachments/assets/1098bf6f-2a5c-40e8-b359-8e96cf68a413)

I've added a warning indicator if the verification grid loses focus (bound to the `blur` event)

I originally tried some text "Shortcuts are disabled" with a tooltip and button to re-gain focus, but there was a lot of cumulative shift, and the elements would move around a lot, so I ended up going with a simple warning icon + tooltip

atruskie commented on src/app/components/annotations/pages/verification/verification.component.ts:
```diff
@@ -129,7 +120,17 @@ class VerificationComponent
+    // use the shortcut keys
+    this.verificationGridElement.nativeElement.focus();
+
+    const timeoutDuration = 1_000 as const;
```
milliseconds?

our conventions are to suffix a unit in the variable name unless the value is a SI unit.

atruskie commented on src/app/components/shared/if-logged-in/if-logged-in.component.ts:
this component is more simple than i was envisioning. I was thinking it could take an input for a capability to test against. that would make it useable for say the `canUpload` or `canDownload` audio capabilities. maybe todo for future scope?

but i'd consider renaming it to allow for future expansion. Maybe `can`. e.g. `<can if-logged-in>` or `<can [capability]="canDownloadCapability">`

hudson-newey replied:
![image](https://github.com/user-attachments/assets/d8cdbddf-c6a1-4dae-ad27-d49aa11ba6e9)

I like the idea. I made an interface that we can implement. It returns a watcher and a predicate that determines if the content should be enabled

atruskie commented on src/app/components/shared/if-logged-in/if-logged-in.component.ts:
```diff
@@ -0,0 +1,45 @@
+    for (const element of content) {
+      // if the disabled attribute is in the elements prototype, we want to
+      // set it to true
+      if ("disabled" in element) {
```
hmmm this is testing the property is present not the attribute?

hudson-newey replied:
I've read the w3 spec, and all `HTMLElement`'s have a global `disabled` attribute, meaning that this check is not necessary

<https://www.w3.org/TR/2011/WD-html5-20110525/elements.html>

atruskie commented on src/app/components/shared/if-logged-in/if-logged-in.component.ts:
```diff
@@ -0,0 +1,45 @@
+  private disableInteractiveContent(state: boolean): void {
+    // TODO: use a ContentChild decorator here
+    const content =
+      this.elementRef.nativeElement.querySelectorAll("button, input");
```
we have other elements that have a disabled attribute. I'm not convinced we should be scanning for sub-elements to set this on. we should just set the attribute on the first child.

atruskie commented on src/app/helpers/context/context-decorators.ts:
```diff
@@ -0,0 +1,39 @@
+        originalNgAfterViewInit();
+      }
+
+      const handler = target[propertyKey].bind(this);
```
does it trigger angular change detection trigger when handler in invoked from the context provider

hudson-newey replied:
No, calling the property doesn't inherently trigger a change detection cycle

I have fixed it so that it requests a change detection cycle after a context has been provided

atruskie commented on src/app/models/AudioRecording.ts:
```diff
@@ -131,6 +131,10 @@ export class AudioRecording
     return apiRoot + audioRecordingOriginalEndpoint(this.id);
   }
 
+  public getSplittableUrl(apiRoot: string): string {
```
we use the term `media` to describe fragments of an audio recording file. 
```suggestion
  public getMediaUrl(apiRoot: string): string {
```

atruskie commented on src/app/services/baw-api/audio-recording/audio-recordings.service.ts:
```diff
@@ -30,6 +33,12 @@ const downloadEndpoint = stringTemplate`/audio_recordings/downloader`;
+ * A path that can be used to download audio recordings that have been split
+ * with start_offset and end_offset url parameters
+ */
+export const audioRecordingSplitableEndpoint = stringTemplate`/audio_recordings/${audioRecordingId}/media.${fileExtension}`;
```
naming s/splitable/media/

hudson-newey committed:
Dynamically import web components from dev server

atruskie commented on src/app/services/models/annotation.service.ts:
```diff
@@ -62,3 +66,62 @@ export class AnnotationService {
   }
 }
+
+interface ResolverNames {
```
custom resolvers are pretty rare right? maybe document why this one is needed (virtual service that need multiple other services)

hudson-newey committed:
Review feedback

hudson-newey committed:
Fix review feedback and fix grid size shrinking bug

hudson-newey committed:
Minor formatting and typing changes

hudson-newey committed:
Bump web components version for merging

hudson-newey committed:
Revert grid size changes

hudson-newey committed:
Disable a faulty test and add todo

hudson-newey committed:
Change to browser-esbuild builder for ssr compatability

hudson-newey committed:
Don't use window in ssr context
